### PR TITLE
[AIRFLOW-5828] Vastly simplify build scripts. Depends on [AIRFLOW-5704] [AIRFLOW-5842]

### PR DESCRIPTION
**Pull request by potiuk**
NOTE FOR REVIEWER: This change depends on #6373 #6499. Please check only last commit.

* Generic Build sript logic is now moved from hooks/build to scripts/ci
* All the scripts are split to smaller functions
* The functions are now grouped in scripts/ci/utils/*.sh by build stage
* The logic of builds is vastly simplified - it's sequence of functions
* The "common" directory is moved to script/ci/utils/
* The DockerHub build-specific logic is now moved to hooks/build
* You can see the progress of build while pre-commit builds (spinner)
* You can see the log file where progress of the build is displayed

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5828

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

### Tests

- [X] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [X] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [X] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### CODE REVIEW

**Comment by ashb** [comment]
@potiuk The chain of (depends on PR x) seems to loop round in a circle.

Could you let me know which PR I should start at/update the notice at the top of the PR(s)?

**Comment by mik-laj** [comment]
@ashb 
```
[AIRFLOW-5704] Improve Kind Kubernetes scripts for local testing …
[AIRFLOW-5842] Switch to debian buster image …
[AIRFLOW-5828] Vastly simplify build scripts …
```

**Comment by potiuk** [comment]
Please don't !

File: scripts/ci/utils/_run.sh [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    _run_ci_script_with_docker "/opt/airflow/scripts/ci/in_container/run_check_licence.sh"
+}
+
+function run_mypy() {
+    FILES=("$@")
+    if [[ "${#FILES[@]}" == "0" ]]; then
+        _run_ci_script_with_docker "/opt/airflow/scripts/ci/in_container/run_mypy.sh" "airflow" "tests" "docs"
+    else
+        _run_ci_script_with_docker "/opt/airflow/scripts/ci/in_container/run_mypy.sh" "$@"
+    fi
```
Comment (ashb) [inline scripts/ci/utils/_run.sh:77]: A trick you can do to remove the similar-almost identical if/else here:

```bash
    if [[ $# -eg 0 ]]; then
        set -- airflow" "tests" "docs"
    fi
    _run_ci_script_with_docker "/opt/airflow/scripts/ci/in_container/run_mypy.sh" "$@"
```

Or if the `set --` is a bit obscure (it sets "$@"/argv), then how about:

```
    FILES=("$@")
    if [[ "${#FILES[@]}" == "0" ]]; then
        FILES=(airflow tests docs)
    fi
    _run_ci_script_with_docker "/opt/airflow/scripts/ci/in_container/run_mypy.sh" "${FILES[@]}"
```

**Comment by dimberman** [comment]
@potiuk is there any way to break this into smaller PRs? Merging 78 files of configuration is giving me anxiety.

**Comment by potiuk** [comment]
It does... I am actually already sneaking in some changes from this PR into other PRs one-by-one :)....  I will definitely try to split it very soon....

**Comment by potiuk** [comment]
Will close it soon as it's been superseded by #7430, #7433 and one more that will come next.

**Comment by potiuk** [comment]
Most of the changes split and merged separately @dimberman - one left that I am going to implement soon and close this one :)

**Comment by potiuk** [comment]
Has been split and implemented in #7430 #7433 #7618

**Pull request closed without merging**